### PR TITLE
WIP: Maintenance Updates for Updated Dependencies

### PR DESCRIPTION
**Pull request by dorukkilitcioglu**
This PR contains a few updates to ensure that we iron out any kinks of TextWiser in regards to the latest library versions. This includes the unnecessary warning message from transformers, the new randomness in UMAP, and finally fixing the numerical instability of SVD.

Plus, the PR adds a way to install the optional dependencies through pip, and also adds weekly CI runs to ensure that if there is a breaking update to one of our dependencies (even the optional ones), we catch it early.

Note that although the outputs of UMAP and SVD are changed compared to previous version, the SVD implementation was not deterministic before, and UMAP is currently non-deterministic if you use the latest version. So I don't consider these changes as "breaking" to warrant a 2.0 version.

### CODE REVIEW

File: CHANGELOG.txt
Comment (skadio) [inline CHANGELOG.txt]: what does "full" requirement mean? add some mention here

File: .github/workflows/ci.yml
Comment (skadio) [inline .github/workflows/ci.yml:10]: what is scheduled?

File: textwiser/embeddings/word.py
Comment (skadio) [inline textwiser/embeddings/word.py:41]: interesting!

Comment (skadio) [reply textwiser/embeddings/word.py:41]: nice history/context logging in comments, btw.

Comment (skadio) [reply textwiser/embeddings/word.py:41]: @dorukkilitcioglu  @takojunior if this is available in TextWiser.. would we still need that code snippet in mab2rec 2_feat_eng.nb

File: textwiser/options.py
Comment (skadio) [inline textwiser/options.py]: is this a tongue-in-cheek comment? :D Use none to  get non-repr results.. is this ever desirable? may be?

File: textwiser/options.py
Comment (skadio) [inline textwiser/options.py]: just checking if you had Constants.default_seed avail here? Either way is fine

File: textwiser/transformations/svd.py
Comment (skadio) [inline textwiser/transformations/svd.py:26]: this is some deep stuff. wonder if it fits within Reproducibility for RecSys'22

File: CHANGELOG.txt
Comment (skadio) [reply CHANGELOG.txt]: oh I see. it points to full_req.txt.. some may be explicitly say full_requirements.txt so that people know where it refers to

File: full_requirements.txt
Comment (skadio) [inline full_requirements.txt]: what's the idea here? Will full_req.txt and req.txt live together? What's their difference and how does one invoke one vs. the other on pip install

**Review by skadio — COMMENTED** [review]
Left minor comments, probably not that useful

File: textwiser/embeddings/word.py
Comment (takojunior) [reply textwiser/embeddings/word.py:41]: Just connected with @dorukkilitcioglu, so this could be the place to remove the logging message in feat_engining notebook from TextWiser side. Then we can drop the snippet in mab2rec.

File: .github/workflows/ci.yml
Comment (dorukkilitcioglu) [reply .github/workflows/ci.yml:10]: Currently the automated tests run when something is pushed to a PR or main branch. By adding this schedule, they will also run every week, even if there are no commits. The aim here is to ensure that even if the underlying libraries release new versions, TextWiser is still installable and tests still pass. If one of the dependencies breaks TW, this would tell us.

File: textwiser/embeddings/word.py
Comment (dorukkilitcioglu) [reply textwiser/embeddings/word.py:41]: Correct, once this is merged we don't need the M2R bit.

File: full_requirements.txt
Comment (dorukkilitcioglu) [reply full_requirements.txt]: These are our optional requirements. When you do `pip install textwiser`, you would install our regular requirements.txt. If you do `pip install textwiser[full]`, it would install both requirements.txt and full_requirements.txt. If you do the full installation you won't have to go and individually install tensorflow for USE, UMAP, etc.

Comment (skadio) [reply full_requirements.txt]: Makes sense, I assume you capture this comment above somewhere (readme? installation page?) so that you user knows about pip install textwiser vs. pip install textwiser[full] 

One follow-up question; I am still missing where/why full installation is needed? What in TextWiser would not run (usage example? tests?) when it is not installed with [full]

File: .github/workflows/ci.yml
Comment (skadio) [reply .github/workflows/ci.yml:10]: awesome! Running in "schedule" seems like a great practice for continuous testing. 

Can you please bring this up with the team, so that one person in each other library can mimic the same? So we not only learn and improve but also scale it out.

Great addition!

File: textwiser/options.py
Comment (dorukkilitcioglu) [reply textwiser/options.py]: I tend to use these non-random options when testing, to quickly try a handful runs and see the range of values I get. But you can also very easily do the same by picking different seeds, so I can be persuaded to remove the optionality.

File: full_requirements.txt
Comment (dorukkilitcioglu) [reply full_requirements.txt]: If you don't use `pip install textwiser[full]`, you cannot use ELMo, USE, UMAP, etc. All of these are in our optional dependencies, in an attempt to make the library "lighter". The library does work without them, it will simply give you an error if you try to initialize a USE model without having tensorflow. Tests also pass without them, those tests are not run if you don't have the corresponding library.

This is in install page, I'll add to readme as well.

File: textwiser/options.py
Comment (skadio) [reply textwiser/options.py]: I see. Good to know. From a "documentation" point of view, it was not clear to why would the user want reproducible results. May be re-word? None returns non-deterministic results?

File: full_requirements.txt
Comment (skadio) [reply full_requirements.txt]: yeah a one-liner in the Readme to expose the availability of [full] and what it does should do it

**Comment by dorukkilitcioglu** [comment]
Hi @skadio @takojunior, after a long while this PR is finally ready to merge. If you have any time to give feedback I'll wait until EOD Tuesday to merge. Thanks!

File: CHANGELOG.txt
Comment (skadio) [inline CHANGELOG.txt:8]: Not a must, just mentioning as a nice-to-have, opportunity, IF easy, linking to code for some (major) updates might be good. Again, not necessary, just mentioning it. For instance, update umap (what file/method is that even if it is not exact line)

File: full_requirements.txt
Comment (skadio) [inline full_requirements.txt]: (minor comment) I would slightly rather this file to be called "requirements_full.txt" so that both files are next to each other in the DIR. The first thing people seek is "reqs" if they want to find it and seeing two reqs file make the intention immediately clear.

File: setup.py
Comment (skadio) [inline setup.py:27]: cool!

File: tests/test_compound.py
Comment (skadio) [inline tests/test_compound.py:45]: I understand the test changes. Shall we expect/account for such behavior change "internally" where textwiser is used?

File: textwiser/embeddings/doc2vec.py
Comment (skadio) [inline textwiser/embeddings/doc2vec.py:27]: nice explanation! Future-developer thanks you

File: textwiser/options.py
Comment (skadio) [inline textwiser/options.py:437]: nice clarififcation on why non-repr might be useful

File: full_requirements.txt
Comment (dorukkilitcioglu) [reply full_requirements.txt]: This makes perfect sense, let me change the filename and we can all pretend it was always named `requirements_full.txt` 😄

File: textwiser/transformations/svd.py
Comment (skadio) [reply textwiser/transformations/svd.py:26]: (we now know that it "fits".. but we will see what the verdict is)

**Review by skadio — APPROVED** [review]
LGTM! This is a nice/critical change, thanks for contributing it. 

I only have one high-level comment/observation which is not specifically for this PR but in general for all of us/team: It is typically good habit to keep PR time-windows narrow. 

When working in real-pace large code bases, the window of oppt for opening and merging PRs is very narrow (like the next morning, or EOD, or even hours at times). Now, we don't work in "that" extreme but building the habit is good for you/us in general. 

The code branches can remain open for a long time and have their own lifespan, that is fine, but when it's ready to go, and only when it's time to go, the PR kicks-off, reviewers make the context switch to the PR, code reviews kick-in, updates follow, and PR is merged (assuming no major revisions or design change). 

Again, it does not apply to this PR (this was more of a maintenance update etc. I am aware of the context here), even then, notice the 27 days between my first review (where I did a full context switch) and second review today (another switch). That's quite a cognitive load to remember what was happening here etc.. (\times everyone reviewing it)

You see my point, and again, only mentioning because it is good skill/habit for you to sharpen in low-key environment. 

Great stuff --thank you !! Open-source does not mean much unless we stay on top of our libraries, and you have shown that ownership repeatedly --way to go!

File: tests/test_compound.py
Comment (dorukkilitcioglu) [reply tests/test_compound.py:45]: Only for pipelines that rely on SVD & UMAP (& possibly Doc2Vec). That being said, before this PR those models would be non-deterministic with latest requirements. So some of the numbers could change after this PR, but if you installed TextWiser before this PR to a different machine, there was a possibility that you wouldn't get the same numbers either.

I think we should be on the lookout for any tests failing, but I don't think we're relying on the exact numbers generated by SVD or UMAP anywhere in the rest of our open source stack.

**Pull request merged**
